### PR TITLE
[FEMDEM & CL Structural] Adding some improvements

### PR DESCRIPTION
**Pull request by AlejandroCornejo**
Adding DEM-FEM contact to the FEMDEM app. Adding some variables in StructuralApp to see the damage in the fiber and matrix in composite CL.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -197,8 +197,8 @@ endif(${MSVC12})
 # Common flags for MSVC
 if(${MSVC})
   message("Detected compiler as MSVC")
-  SET (CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /W1 /bigobj /EHsc -DBOOST_ALL_NO_LIB -D_SCL_SECURE_NO_WARNINGS")
-  SET (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /W1 /bigobj /EHsc -DBOOST_ALL_NO_LIB -D_SCL_SECURE_NO_WARNINGS")
+  SET (CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /W1 /bigobj /EHsc -DBOOST_ALL_NO_LIB -D_SCL_SECURE_NO_WARNINGS -DNOMINMAX")
```
Comment (loumalouomega) [inline CMakeLists.txt]: @roigcarlo check this

File: applications/FemToDemApplication/custom_processes/generate_dem_process.cpp [hunk 2]
```diff
@@ -168,4 +181,20 @@ int GenerateDemProcess::GetMaximumDEMId()
...
 }
 
+/***********************************************************************************/
+/***********************************************************************************/
+
+int GenerateDemProcess::GetMaximumFEMId()
+{
+    int max_id = 0;
+    const auto it_FEM_node_begin = mrModelPart.NodesBegin();
+    // #pragma omp parallel for
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/generate_dem_process.cpp]: you can look in mortar_utilities for a OMP code to get the max and min

File: applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp [hunk 3]
```diff
@@ -51,6 +51,7 @@ void UpdateDemKinematicsProcess::UpdateKinematics(
 {
     const array_1d<double,3> coordinates = rNode->Coordinates();
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp:52]: Add references if needed to avoid copies

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Some minor comments,a nd waiting for the review of  @roigcarlo related with the change affecting to the CMakeLists.txt

File: applications/StructuralMechanicsApplication/structural_mechanics_application_variables.h [hunk 4]
```diff
@@ -39,7 +39,7 @@ namespace Kratos
 
     enum class SofteningType {Linear = 0, Exponential = 1};
 
-    enum class TangentOperatorEstimation {Analytic = 0, FirstOrderPerturbation = 1, SecondOrderPerturbation = 2};
+    enum class TangentOperatorEstimation {Analytic = 0, FirstOrderPerturbation = 1, SecondOrderPerturbation = 2, Secant = 3};
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/structural_mechanics_application_variables.h:42]: FYI @KratosMultiphysics/structural-mechanics

File: CMakeLists.txt (hunk 1 above)
Comment (philbucher) [reply CMakeLists.txt]: This should no longer be necessary, we solved it globally I think

Comment (AlejandroCornejo) [reply CMakeLists.txt]: I'll remove that, THX

Comment (roigcarlo) [reply CMakeLists.txt]: Fine to me, as @philbucher says that's already solved

**Comment by philbucher** [comment]
general question, why don't you put the Gid-interface to the GiD-interface repo?

File: applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp (hunk 3 above)
Comment (AlejandroCornejo) [reply applications/FemToDemApplication/custom_processes/update_dem_kinematics_process.cpp:52]: I was tempted to do that but I noticed that in some cases is kind of dangerous when crossing refs...

File: CMakeLists.txt (hunk 1 above)
Comment (AlejandroCornejo) [reply CMakeLists.txt]: done

File: applications/FemToDemApplication/custom_processes/generate_dem_process.cpp (hunk 2 above)
Comment (AlejandroCornejo) [reply applications/FemToDemApplication/custom_processes/generate_dem_process.cpp]: I couldn't find the methods @loumalouomega  :S

**Comment by AlejandroCornejo** [comment]
> 
> 
> general question, why don't you put the Gid-interface to the GiD-interface repo?

I understand the question and I agree with you. The thing is that when I started I used the Solid Ptype but at that time the changes were too much (in terms of number and in terms of periodicity)  and this is why I created mine.
At some point I'll migrate to the Structural (I've done a lot fo changes to do that) and I'll use the Structural ptype as a base. :)

File: applications/FemToDemApplication/custom_processes/generate_dem_process.cpp (hunk 2 above)
Comment (loumalouomega) [reply applications/FemToDemApplication/custom_processes/generate_dem_process.cpp]: Sorry, here: https://github.com/KratosMultiphysics/Kratos/blob/80128a441780b58b7389563bf6d286a7b2a3ffe6/applications/ContactStructuralMechanicsApplication/custom_utilities/contact_utilities.h#L111

**Comment by AlejandroCornejo** [comment]
ping @loumalouomega

File: applications/FemToDemApplication/custom_processes/generate_initial_skin_DEM_process.cpp [hunk 5]
```diff
@@ -149,4 +162,20 @@ int GenerateInitialSkinDEMProcess::GetMaximumDEMId()
...
 
+/***********************************************************************************/
+/***********************************************************************************/
+
+int GenerateInitialSkinDEMProcess::GetMaximumFEMId()
+{
+    int max_id = 0;
+    const auto it_FEM_node_begin = mrModelPart.NodesBegin();
+    // #pragma omp parallel for
+    for (int i = 0; i < static_cast<int>(mrModelPart.Nodes().size()); i++) {
```
Comment (loumalouomega) [inline applications/FemToDemApplication/custom_processes/generate_initial_skin_DEM_process.cpp:181]: Still missing OMP max and min

**Review by loumalouomega — COMMENTED** [review]
Comment

**Review by loumalouomega — APPROVED** [review]
Now

File: applications/DemStructuresCouplingApplication/custom_python/add_custom_utilities_to_python.cpp [hunk 6]
```diff
@@ -60,7 +60,7 @@ namespace Kratos {
                 .def("ExecuteFinalizeSolutionStep", &StressFailureCheckUtilities::ExecuteFinalizeSolutionStep)
             ;
 
-            class_<PostProcessUtilities, PostProcessUtilities::Pointer>(m, "PostProcessUtilities")
+            class_<PostProcessUtilities,PostProcessUtilities::Pointer>(m, "PostProcessUtilities", module_local())
```
Comment (philbucher) [inline applications/DemStructuresCouplingApplication/custom_python/add_custom_utilities_to_python.cpp:63]: what does this do?

Comment (AlejandroCornejo) [reply applications/DemStructuresCouplingApplication/custom_python/add_custom_utilities_to_python.cpp:63]: Hi @philbucher. This is a temporal solution that @maceligueta and me have found for a problem regarding the PostProcessUtilities. Without adding this, the code breaks... Maybe @maceligueta can add some info

Comment (maceligueta) [reply applications/DemStructuresCouplingApplication/custom_python/add_custom_utilities_to_python.cpp:63]: Hi @philbucher , this is an important matter that I was planning to post on a new issue. Basically, when you have utilities with the same name in different applications you may get the following error:
```
from KratosDemStructuresCouplingApplication import *
ImportError: generic_type: type "PostProcessUtilities" is already registered!
```
This is kind of mind-blowing, because the examples we tried were not doing any 'import *' and the utilities were clearly encapsulated in each respective application. Also @JoseMGonzalez-UPC  experienced this in his private branch for drilling-beam-application. We requested @roigcarlo 's help and finally this solution worked. 
This made me think that we will come across with this issue frequently in the future, when people develop more and more couplings between applications.
My proposal is to add 'module_local()' in all applications by default. 
Maybe @roigcarlo can bring his opinion to clarify.

File: applications/FemToDemApplication/custom_processes/generate_initial_skin_DEM_process.cpp (hunk 5 above)
Comment (rubenzorrilla) [reply applications/FemToDemApplication/custom_processes/generate_initial_skin_DEM_process.cpp:181]: This could be a candidate to be implemented in  #943

Comment (loumalouomega) [reply applications/FemToDemApplication/custom_processes/generate_initial_skin_DEM_process.cpp:181]: The mytical #943

File: applications/DemStructuresCouplingApplication/custom_python/add_custom_utilities_to_python.cpp (hunk 6 above)
Comment (philbucher) [reply applications/DemStructuresCouplingApplication/custom_python/add_custom_utilities_to_python.cpp:63]: Hm this is interesting

so lets say util `Abc` exists in `XApplication` and `YApplication`, then the following would still fail?
~~~py
from XApplication import Abc as x_abc
from YApplication import Abc as x_abc
~~~

and if you add `module_local` then it works?

Comment (maceligueta) [reply applications/DemStructuresCouplingApplication/custom_python/add_custom_utilities_to_python.cpp:63]: Well, I guess this example is weirder because you are overwriting x_abc.
The problem we detected was raised at
```
import XApplication as X
import YApplication as Y  #<--- ERROR
```
where both applications have an utility called `Abc`

Comment (philbucher) [reply applications/DemStructuresCouplingApplication/custom_python/add_custom_utilities_to_python.cpp:63]: > Well, I guess this example is weirder because you are overwriting x_abc.

ok I did not intend that, was typo

But thanks for clarifying

adding it to the todo of the committee

Comment (maceligueta) [reply applications/DemStructuresCouplingApplication/custom_python/add_custom_utilities_to_python.cpp:63]: I just wanted to make clear that, even though in the example I posted I wrote
```
from KratosDemStructuresCouplingApplication import *
ImportError: generic_type: type "PostProcessUtilities" is already registered!
```
with @roigcarlo  and @JoseMGonzalez-UPC  we tried the standard import
`import KratosDemStructuresCouplingApplication as DemStruct`
and the problem was still there.

**Comment by roigcarlo** [comment]
We have been discussing this a bit with the @KratosMultiphysics/technical-committee. Some conclusions/findings:

https://pybind11.readthedocs.io/en/stable/advanced/classes.html#module-local-class-bindings

Apparently, the module_local is not the default behavior as it interferes with the inheritance mechanism, so one cannot for example return a base class as python would still see it as instance of the derived module, not the common one.

Since the PostProcess utilities does not derive from any parent class, the solution works in our case, but it is important to notice that we cannot use module_local everywhere. One of the cases we should avoid is as follows:


```C++
// Module BASE
class PostProcess {
}

class UserCalss {
    void UserFunction(PostProcess &p) {
        p.write();
    }
}

class_<PostProcess,PostProcess::Pointer>(m, "PostProcess")
class_<UserCalss, UserCalss::Pointer>(m, "UserCalss").def("UserFunction", &UserFunction)
```

```C++
// Module A
class MyPostProcess :  public PostProcess {
    void write() {...}
}

class_<MyPostProcess, MyPostProcess::Pointer, PostProcess>(m, "MyPostProcess")
```

```C++
// Module B
class MyPostProcess:  public PostProcess {
    void write() {...}
}

class_<MyPostProcess, MyPostProcess::Pointer, PostProcess>(m, "MyPostProcess")
```

In this case, Module A and B which both are exporting a `MyPostProcess` class will colide, and hence they need the `module_local`. But once `module_local` is set, from python point of view `MyPostProcess` is no longer a `PostProcess`  but a `A.PostProcess`, hance no longen being passable for example to `UserCalss.UserFunction`

**TL;DR**

We can use the `module_local` as long as we don't have inheritance on the classes using them, but we cannot use it as default in Kratos.
